php_phongo.c

```diff
@@ -1450,42 +1502,104 @@ static mongoc_client_t *php_phongo_make_mongo_client(const mongoc_uri_t *uri, zv
 {
-	mongoc_uri_t             *uri;
+	char             *hash = NULL;
+	size_t            hash_len;
+	bson_t            bson_options = BSON_INITIALIZER;
+	mongoc_uri_t     *uri = NULL;
+	mongoc_ssl_opt_t *ssl_opt = NULL;
+
+#if PHP_VERSION_ID >= 70000
+	zval             *client_ptr;
+	zval              new_client_ptr;
+#else
+	mongoc_client_t **client_ptr;
+#endif
 
-	if (!(uri = php_phongo_make_uri(uri_string, bson_options))) {
+	if (!(hash = php_phongo_manager_make_client_hash(uri_string, options, driverOptions, &hash_len TSRMLS_CC))) {
+		/* Exception should already have been thrown and there is nothing to free */
+		return;
+	}
+
+#if PHP_VERSION_ID >= 70000
+	if ((client_ptr = zend_hash_str_find(&MONGODB_G(clients), hash, hash_len)) && Z_TYPE_P(client_ptr) == IS_PTR) {
+		MONGOC_DEBUG("Found client for hash: %s\n", hash);
+		manager->client = (mongoc_client_t *)Z_PTR_P(client_ptr);
+		goto cleanup;
+	}
+#else
+	if (zend_hash_find(&MONGODB_G(clients), hash, hash_len + 1, (void**) &client_ptr) == SUCCESS) {
+		MONGOC_DEBUG("Found client for hash: %s\n", hash);
+		manager->client = *client_ptr;
+		goto cleanup;
+	}
+#endif
+
+	if (options) {
+		phongo_zval_to_bson(options, PHONGO_BSON_NONE, &bson_options, NULL TSRMLS_CC);
+	}
+
+	/* An exception may be thrown during BSON conversion */
+	if (EG(exception)) {
+		goto cleanup;
+	}
+
+	if (!(uri = php_phongo_make_uri(uri_string, &bson_options))) {
 		phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Failed to parse MongoDB URI: '%s'", uri_string);
-		return false;
+		goto cleanup;
 	}
 
-	if (!php_phongo_apply_rc_options_to_uri(uri, bson_options TSRMLS_CC) ||
-	    !php_phongo_apply_rp_options_to_uri(uri, bson_options TSRMLS_CC) ||
-	    !php_phongo_apply_wc_options_to_uri(uri, bson_options TSRMLS_CC)) {
+	if (!php_phongo_apply_rc_options_to_uri(uri, &bson_options TSRMLS_CC) ||
+	    !php_phongo_apply_rp_options_to_uri(uri, &bson_options TSRMLS_CC) ||
+	    !php_phongo_apply_wc_options_to_uri(uri, &bson_options TSRMLS_CC)) {
 		/* Exception should already have been thrown */
-		mongoc_uri_destroy(uri);
-		return false;
+		goto cleanup;
 	}
 
-	manager->client = php_phongo_make_mongo_client(uri, driverOptions TSRMLS_CC);
-	mongoc_uri_destroy(uri);
+	ssl_opt = php_phongo_make_ssl_opt(driverOptions TSRMLS_CC);
+
+	/* An exception may be thrown during SSL option creation */
+	if (EG(exception)) {
+		goto cleanup;
+	}
+
+	manager->client = php_phongo_make_mongo_client(uri, ssl_opt TSRMLS_CC);
 
 	if (!manager->client) {
 		phongo_throw_exception(PHONGO_ERROR_RUNTIME TSRMLS_CC, "Failed to create Manager from URI: '%s'", uri_string);
-		return false;
+		goto cleanup;
 	}
 
-	return true;
+	MONGOC_DEBUG("Created client hash: %s\n", hash);
+#if PHP_VERSION_ID >= 70000
+	ZVAL_PTR(&new_client_ptr, manager->client);
+	zend_hash_str_update(&MONGODB_G(clients), hash, hash_len, &new_client_ptr);
```

Comment:
Travis says that hash_len might be able to use uninitialized: https://travis-ci.org/mongodb/mongo-php-driver/jobs/158304272#L1220